test/unit/math/test_ad_matvar.hpp

```diff
@@ -0,0 +1,724 @@
+  FAIL() << "test_matvar_gradient should only be used when the return is a "
+            "reverse mode autodiff type";
+}
+///@}
+
+/**
+ * Given the inputs and outputs of a function called with Eigen
+ * matrices of vars (matvars) and vars with inner Eigen types (varmats),
+ * check that the values of the outputs are equal and that gradients with
+ * respect to each output are the same
+ *
+ * This overload works when the outputs are `std::vector` types.
+ *
+ * @tparam ResultMatVar Output type of matvar calculation
+ * @tparam ResultVarMat Output type of varmat calculation
+ * @tparam MatVar Argument type for matvar calculation
+ * @tparam VarMat Argument type for varmat calculation
+ * @param tols Test tolerances
+ * @param A_mv_f Output for matvar calculation
+ * @param A_vm_f Output for varmat calculation
+ * @param A_mv Input argument to matvar calculation
+ * @param A_vm Input argument to varmat calculation
+ */
+template <typename ResultMatVar, typename ResultVarMat, typename MatVar,
+          typename VarMat,
+          require_all_std_vector_t<ResultMatVar, ResultVarMat>* = nullptr>
+inline void test_matvar_gradient(const ad_tolerances& tols,
+                                 ResultMatVar& A_mv_f, ResultVarMat& A_vm_f,
+                                 const MatVar& A_mv, const VarMat& A_vm) {
+  for (size_t i = 0; i < A_vm_f.size(); ++i) {
+    test_matvar_gradient(tols, A_mv_f[i], A_vm_f[i], A_mv, A_vm);
+  }
+}
+
+/**
+ * Check that if at least one of the input types is a var with inner
+ * Eigen type that the output is not an Eigen matrix of vars.
+ *
+ * Similarly check that if none of the inputs are vars with inner type
+ * Eigen that the output is not a var with inner Eigen type.
+ *
+ * @tparam ReturnType Type of result of calculation
+ * @tparam Types Types of input arguments
+ * @param ret not used, only types are used.
+ * @param x not used, only types are used.
+ */
+template <typename ReturnType, typename... Types,
+          require_all_not_std_vector_t<ReturnType, Types...>* = nullptr,
+          require_any_matrix_t<Types...>* = nullptr>
+void check_return_type(const ReturnType& ret, const Types&... x) {
+  using stan::is_eigen;
+  using stan::is_var_matrix;
+  using stan::math::var_value;
+  using stan::math::test::type_name;
+  if (stan::math::conjunction<is_eigen<Types>...>::value
+      && !(is_eigen<ReturnType>::value
+           || std::is_same<ReturnType, var_value<double>>::value)) {
+    FAIL() << "All the arguments are `Eigen::Matrix<T, R, C>` types, but the "
+              "return type is neither an `Eigen::Matrix<T, R2, C2>` or a `var`";
+  } else if (stan::math::disjunction<is_var_matrix<Types>...>::value
+             && !(is_var_matrix<ReturnType>::value
+                  || std::is_same<ReturnType, var_value<double>>::value)) {
+    FAIL() << "One of the arguments is a `var_value<Eigen::Matrix<double, R, "
+              "C>>`, but the return type is not a `var_value<T>`";
+  }
```

Comment:
Will this work with vectors of mat var types?

Reply:
No. Everything that uses `std::vector` gets shoveled through [this check](https://github.com/stan-dev/math/pull/2280/files/70766a8c81fb652478c8c09af4e42c94db4d89b1#diff-7dc2c9d1d99265ed85e16bd88a0541f7b79e39ea27a68bc27a83578be76e7b26R315)

Our `std::vector` support is limited. I think next step here would be to bring the `std::vector` support in, but I didn't do it cuz we've been building this testing in pieces (only what we need).